include/tnt/linear/impl/Gaussian_elimination_impl.hpp

```diff
@@ -0,0 +1,249 @@
+#ifndef TNT_GAUSSIAN_ELIMINATION_IMPL_HPP
+#define TNT_GAUSSIAN_ELIMINATION_IMPL_HPP
+
+#include <tnt/utils/testing.hpp>
+#include <tnt/utils/simd.hpp>
+#include <tnt/linear/Gaussian_elimination.hpp>
+
+namespace tnt {
+
+    namespace detail {
+
+        template<typename DataType>
+
+        struct OptimizedGaussianElimination<DataType> {
+
+            /* returns vector solution to G_E of tensor. Must be of same data type */
+            static Tensor<DataType> eval(Tensor<DataType> &tensor, Tensor<DataType> &vector) {
+
+                DataType *ptr = tensor.data.data;
+                DataType *vct = vector.data.data;
+                int num_rows = tensor.shape.axes[0];
+                int num_cols = tensor.shape.axes[1];
+
+                for (int i = 0; i < num_rows; i++) {
+                    int lead = num_cols * i;
+                    while (ptr[lead] == 0 && lead < (i*num_cols + num_rows)) {
+                        lead++;
+                    }
+
+                    if (ptr[lead] == 1) {
+                        clearColumn(ptr, i, (lead - (num_cols * i)), num_rows, num_cols, vct);
+                    } else if (lead != 0) {
+                        divideRow(ptr, ptr[lead], i, num_cols);
+                        divideVector(ptr, vector, i);
+                        clearColumn(ptr, i, (lead - (num_cols * i)), num_rows, num_cols, vct);
+                    }
+                }
+
+                naturalOrder(ptr, num_cols, num_rows, vct);
+                vector.data = AlignedPtr<DataType>(vct, vector.shape.total());
+
+                return vector;
+            }
+
+            static Tensor<DataType> eval(Tensor<DataType> &tensor) {
+
+                DataType *ptr = tensor.data.data;
+                int num_rows = tensor.shape.axes[0];
+                int num_cols = tensor.shape.axes[1];
+
+                for (int i = 0; i < num_rows; i++) {
+                    int lead = num_cols * i;
+                    while (ptr[lead] == 0 && lead < (i*num_cols + num_rows)) {
+                        lead++;
+                    }
+
+                    if (ptr[lead] == 1) {
+                        clearColumn(ptr, i, (lead - (num_cols * i)), num_rows, num_cols, nullptr);
+                    } else if (lead != 0) {
+                        divideRow(ptr, ptr[lead], i, num_cols);
+                        clearColumn(ptr, i, (lead - (num_cols * i)), num_rows, num_cols, nullptr);
+                    }
+                }
+                naturalOrder(ptr, num_cols, num_rows, nullptr);
+                tensor.data = AlignedPtr<DataType>(ptr, tensor.shape.total());
+
+                return tensor;
+            }
+
+        private:
+
+            static void divideVector(DataType* vct, int row, DataType value)
+            {
+                vct[row] = vct[row] - value;
```

Comment:
Is it expected that vct is updated in place? Meaning if I pass in 2 tensors should I expect both to change?